lua/entities/gmod_wire_expression2/core/table.lua

```diff
@@ -541,32 +541,59 @@ end
 
 __e2setcost(10)
 
-local function clone(self, tbl, lookup)
+--- deepcopy(self, t) creates a clone of t.
+-- All of its fields are also copied.
+local function deepcopy(self, tbl, lookup)
+	self.prf = self.prf + opcost * 2 -- creating new table
 	local copy = {}
 
 	lookup = lookup or {}
 	lookup[tbl] = copy
 
 	for k, v in pairs(tbl) do
+		self.prf = self.prf + opcost
 		if type(v) == "table" then
-			if lookup[v] then
-				self.prf = self.prf + opcost -- simple assign operation
-				copy[k] = lookup[v]
-			else
-				self.prf = self.prf + opcost * 3 -- creating new table
-				copy[k] = clone(self, v, lookup)
-			end
+			copy[k] = lookup[v] or deepcopy(self, v, lookup)
 		else
-			self.prf = self.prf + opcost -- simple assign operation
 			copy[k] = v
 		end
 	end
+	return copy
+end
+
+--- cloneTable(self, t) creates a clone of t, where t is an E2 table.
+-- Its fields are copied if they need copying (which depends on the )
+local function cloneTable(self, tbl, lookup)
+	self.prf = self.prf + opcost * 2 -- creating new table
+	local copy = { size = tbl.size }
+
+	lookup = lookup or {}
+	lookup[tbl] = copy
+
+	local function cloneTablePart(types, values)
+		local newTypes, newValues = {}, {}
+		for name, e2type in pairs(types) do
+			newTypes[name] = e2type
+			local value = values[name]
+			self.prf = self.prf + opcost
+			if e2type == "t" then -- the field is also an E2 table
+				newValues[name] = lookup[value] or cloneTable(self, value, lookup)
+			elseif tbls[e2type] then -- the field is some other mutable type
```

Comment:
Vectors are mutable, so we may need to add a special case for them here.

Reply:
yes, you'll need to do this probably
```
local mutables = {v=true,q=true,...}
[...]
elseif mutables[e2type] then -- do something
```

Reply:
Also I just noticed, we'll have to do a special copy for vectors considering some of them will be tables and some will be userdata.

Reply:
no the whole point was that we don't have to copy tables/angles/quats/etc because they are immutable

Reply:
You can edit vectors and quaternions via the index operator, I guess if you want to say `table:clone()` is only going to clone `tables` and makes no promises about the objects held within them then this is fine.

Reply:
haha you're damn right I'm not!

Reply:
There's special handling for mutable types elsewhere that isn't applied to vectors and quaternions, which I think indicates they were never supposed to be mutable. For example:

```PHP
@inputs In
@outputs Array:array Table:table Vector:vector
Array[1, number] = Table[1, number] = Vector[1] = In
```

In this case, the outputs Array and Table will update when In does, but Vector won't.

Reply:
Hah, @TomyLobo and I have been jointly pissed off about this before in #1348.

Reply:
Yeah since it's too late to fix it... we have a bunch of properties for each type. Let's add another one that marks a type as immutable (with mutable being the default, for safety reasons)

Another idea would be to just use the existing output+input transformation functions for each type in order to obtain a copy.
I think those should handle detangling vectors.
If not, that'd need a fix anyway :)

Reply:
The first idea would work fine I believe.

> In response to @bigdogmat's objection: how about table:deepClone() or table:shallowClone() (depending on the current semantics... i believe table:clone() is deep?)

It was more so an objection to if the behavior was considered correct even if types were mutable (which they are), though it was just an oversight on @AbigailBuccaneer's part.

None the less, I wouldn't be against a `table:shallowClone()` and making `table:deepClone()` an alias of `table:clone()`.